### DISCUSSION TITLE
Can we import the Meta Component Builder function?

### DISCUSSION THREAD

**Discussion by renomureza** [remix-run/remix#4551] [Q&A]
I'm trying to avoid exporting the function `export const meta: MetaFunction=() => {}` on every page, I've managed to make it for JSON-LD and canonical.

All SEO data, returned by the `loader` function of each page.

Does Remix expose the meta component builder function so I can use it anywhere like my `jsonldRender()` function?

`app/root.tsx`
```tsx
import { Seo } from "./components/seo-utils";

export default function App() {
  return (
    <html lang="en" className="h-full">
      <head>
      
        <Seo /> {/* 👋 custom */}
       
      </head>
      <body className="h-full">
         {/* ... */}
      </body>
    </html>
  );
}
```
`app/components/seo-utils/index.tsx`
```tsx
export const Seo = () => {
  const matches = useMatches();

  const result = matches.flatMap((match, i) => {
    const { data } = match;
    const tmpResult = [];

    if (hasSeoData(data)) {
      tmpResult.push(jsonldRender(data.seo.jsonld));
      tmpResult.push(
        linkRender([{ rel: "canonical", href: data.seo.canonical }])
      );
      // tmpResult.push(
      //   meta(data.seo.meta) // 🚨 meta()
      // )
    }

    return tmpResult;
  });

  return result as unknown as ReactElement<any, any>;
};

```

`app/routes/index.tsx`
```tsx
function getLoaderData() {
  return {
    seo: generateSeo({
      path: "/",
      description: site.description,
      title: `${site.name} - ${site.slogan}`,
      breadcrumbItems: [{ title: "Home", path: "/" }],
    }),
  };
}

type LoaderData = ReturnType<typeof getLoaderData>;

export const loader: LoaderFunction = () => {
  return json<LoaderData>(getLoaderData());
};

export default function Index() {
  return ({/* ... */} )
}
```

**Comment by sergiodxa** [comment]
You can build your own MetaFunction with useMatches hook, for JSON-LD check [Remix Util's StructuredDataFunction](https://github.com/sergiodxa/remix-utils#structureddata), you can see the code to see how it's implemented.

**Comment by rodbs** [comment]
@sergiodxa  
Two questions: 
 - When I destructure `parentsData` I get an array of objects (I guess one element per route). But this is not very easy to iterate to look for our data, right?  The `parentsData` in meta function seems much easier (it
 - Does this `StructuredData` work with meta@v2 ? do we need to export a meta for normal SEO tags and StructuredData for json-ld? 
 Thx
